### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] Does not flag on unnecessary nonnull assertion on `declare`d value

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#8794]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.3&fileType=.ts&code=CYUwxgNghgTiAEYD2A7AzgF3gTwFzxQFcBbAIxBgG4AoZdJCEAOgiQHMAKbAQgEpKgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
declare const y: number;
console.log(y!);
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

I expect the second line to report an error.

### Actual Result

No error.

### Additional Info
The snippet
```typescript
const y: number = 3;
console.log(y!);
```
flags as expected. For some reason just `declare` seems to mess it up.

**Comment by bradzacher** [comment]
This likely occurs because the rule has logic to determine if the non-null assertion is a "definite assignment" assertion.

Variables can be declared without a nullish type and not be assigned a value. When we inspect the type at the usage it is a non-nullish type. So a non-null assertion *looks* unnecessary, but it is actually necassery (removing it would be a type error).

**Comment by kirkwaiblinger** [comment]
@bradzacher 
> This likely occurs because the rule has logic to determine if the non-null assertion is a "definite assignment" assertion.
> 
> Variables can be declared without a nullish type and not be assigned a value. When we inspect the type at the usage it is a non-nullish type. So a non-null assertion _looks_ unnecessary, but it is actually necassery (removing it would be a type error).

Oh, fascinating...... and a quick look in the code shows also that you've thought about this before, too :) https://github.com/Microsoft/TypeScript/issues/31124

Just to be clear, though, which of the following do you mean by your response?
1. This behavior is a necessary tradeoff in order to handle `let x!: number`-type cases correctly
1. This behavior is a fixable bug, that is caused by incorrect logic designed to handle `let x!: number`-type cases
1. This behavior may or may not be fixable, investigation would be needed. A good starting point would be to look at the code that is designed to handle the `let x!: number`-type cases

**Comment by bradzacher** [comment]
Cases like this is what I'm referring to
```ts
let x: number;
x!.toFixed();
```
[playground](https://www.typescriptlang.org/play?isolatedModules=true&useUnknownInCatchVariables=true#code/DYUwLgBAHgXBB2BXAtgIxAJwNwCgoEIA6MAewDEBLKEAEwAoBKXIA)

If you remove the non-null assertion then you get a TS error. This sort of code looks bad but there's usually a callback function between the assignment which TS's can't track.

**Comment by kirkwaiblinger** [comment]
Gotcha, yeah, so that seems clear to me, since that's literally a (tricky) case of a necessary nonnull assertion. Do you think, though, that that pattern apply to variables declared with the `declare` keyword? I can't quite picture how it would translate.

**Comment by bradzacher** [comment]
Based on some quick testing - no it does not apply to `declare`'d variables because those variables have no "potentially unassigned" state.
